### ISSUE TITLE
Allow an EmbeddedCollectionField to be updated by a Map data structure.

### ISSUE THREAD

**Issue by esb** [foundryvtt/foundryvtt#7706]
### What happened?

When using the "fromDropData" API call, the call returns an Item with an "effects" field populated as a Map. This is correct as the schema defines effects as "{Collection<BaseActiveEffect>}".

However, when using the returned data to update the object, the warning "Cannot initialize the effects EmbeddedCollection from source data which is not an Array" is issued.

The problem can be resolved by passing the data through a duplicate, which implicitly converts the Map to an Array, but this is a bit obscure.

Shouldn't updated EmbeddedCollections accept a Map as well as an Array, especially as the Map is being returned from other API calls?

### What ways of accessing Foundry can you encounter this issue in?

- [X] Native App (Electron)
- [ ] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Other

### Reproduction Steps

This can be reproduced by using a game system that accepts the "fromDropData" without additionally modifying it with a duplicate or other processing.

### What core version are you reporting this for?

Version 9 (build 269)

### Bug Checklist

- [X] The issue occurs while all Modules are disabled

**Comment by Fyorl** [comment]
This is definitely working as intended. The initialised `DataModel`/`Document` that you work with is asymmetrical to the underlying source data that is persisted to disk. Classes and `Collection`s are instantiated from that source data, but when you update it, you must update the source data directly, rather than using the richer classes.

I will update the description to reframe it as a feature request for accepting `Map`s and `Set`s in updates.

**Comment by Fyorl** [comment]
From @arbron:

Currently in order to perform an update to a set defined as a SetField in a DataModel it must first be converted to an array:

```js
const itemSet = this.document.system.items;
itemSet.add(item.uuid);
this.document.update({"system.items": Array.from(itemSet)});
```

It would be convenient if a set could be passed in:

```js
const itemSet = this.document.system.items;
itemSet.add(item.uuid);
this.document.update({"system.items": itemSet});
```

**Comment by aaclayton** [comment]
This will have to be an improvement for the future, unfortunately it won't make it in time for V10 stable.

**Comment by JPMeehan** [comment]
This seems partially resolved by #10418
